dask/dataframe/shuffle.py

```diff
@@ -24,6 +25,126 @@
 logger = logging.getLogger(__name__)
 
 
+class ShuffleStage(Layer):
+    """Shuffle HLG Layer
+
+    High-level graph layer corresponding to a single stage of
+    an inter-partition shuffle operation.
+
+
+    Stage: (shuffle-group) -> (shuffle-split) -> (shuffle-join)
+
+
+    """
+
+    def __init__(
+        self,
+        ddf,
+        name,
+        column,
+        inputs,
+        stage,
+        npartitions,
+        n,
+        k,
+        ignore_index,
+        parts_out=None,
+    ):
+        self._ddf = ddf
+        self.column = column
+        self.name = name
+        self.inputs = inputs
+        self.stage = stage
+        self.npartitions = npartitions
+        self.n = n
+        self.k = k
+        self.ignore_index = ignore_index
+        self.parts_out = parts_out or range(len(inputs))
+
+    def __repr__(self):
+        return "ShuffleStage<name='{}', stage={}, k={}, npartitions={}>".format(
+            self.name, self.stage, self.k, self.npartitions
+        )
+
+    @property
+    def _dict(self):
+        """Materialize full dict representation"""
+        if hasattr(self, "_cached_dict"):
+            return self._cached_dict["dsk"]
+        else:
+            dsk = _rbc_tasks_stage(
+                self._ddf,
+                self.name,
+                self.column,
+                self.inputs,
+                self.parts_out,
+                self.stage,
+                self.npartitions,
+                self.n,
+                self.k,
+                self.ignore_index,
+            )
+            self._cached_dict = {"dsk": dsk}
+        return self._cached_dict["dsk"]
+
+    def __getitem__(self, key):
+        return self._dict[key]
+
+    def __iter__(self):
+        return iter(self._dict)
+
+    def __len__(self):
+        return len(self._dict)
```

Comment:
Do you have any thoughts on how easy it would be to generate `keys` and `get_dependencies` without generating the full graph?  These are likely to be the next useful steps here in terms of improving performance.

Reply:
Sorry for the delayed response here!  I'm not sure I fully understand the question,  because I still don't really understand when `keys` (which I assume is gerated with `__iter__`) and `get_dependencies` are actually *used* (or needed).

For example, it doesn't seem to me like there is ever a need to call `get_dependencies` on a `ShuffleStage` Layer, so I'm not really sure how to explore that idea.  Also, it is my understanding that the "dunder" methods you highlighted are only needed to operate on the layer after it already needs to be converted to a dictionary.   I am well aware that I have gaps in my understanding here, so feel free to set me straight :)